File: dotnet/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/Services/DocumentTemplateManager.cs
using PrototypeDesignPattern.Example1.Models.Abstract;
using PrototypeDesignPattern.Example1.Models.Concrete;

namespace PrototypeDesignPattern.Example1.Services;

// Prototype Registry: stores pre-configured document templates and produces clones on demand
public class DocumentTemplateManager
{
    private readonly Dictionary<string, DocumentBase> _templates = new();

    public DocumentTemplateManager()
    {
        InitializeTemplates();
    }

    private void InitializeTemplates()
    {
        var cvTemplate = new CVDocument
        {
            Title = "Standard CV Template",
            PersonalInfo = new PersonalInfo
            {
                FullName = "[Your Name]",
                Email = "[your.email@example.com]",
                Phone = "[Your Phone]",
                Address = "[Your Address]"
            }
        };
        cvTemplate.Skills.AddRange(new[] { "[Skill 1]", "[Skill 2]", "[Skill 3]" });
        cvTemplate.Education.Add(new Education
        {
            Institution = "[University Name]",
            Degree = "[Degree]",
            FieldOfStudy = "[Field of Study]",
            GraduationYear = DateTime.Now.Year
        });

        var proposalTemplate = new ProposalDocument
        {
            Title = "Standard Business Proposal",
            ClientInfo = new ClientInfo
            {
                CompanyName = "[Client Company]",
                ContactPerson = "[Contact Person]",
                Email = "[client@example.com]"
            },
            Terms = "Standard terms and conditions apply.",
            ValidityDays = 30
        };
        proposalTemplate.ProposalItems.Add(new ProposalItem
        {
            Description = "[Service/Product Description]",
            Quantity = 1,
            UnitPrice = 0
        });

        var reportTemplate = new ReportDocument
        {
            Title = "Standard Report Template",
            ReportType = ReportType.Management,
            Header = new ReportHeader
            {
                Title = "[Report Title]",
                PreparedBy = "[Your Name]",
                Department = "[Department]"
            }
        };
        reportTemplate.Sections.Add(new ReportSection
        {
            SectionTitle = "Executive Summary",
            Content = "[Executive summary content]",
            Order = 1
        });

        RegisterTemplate("standard-cv", cvTemplate);
        RegisterTemplate("standard-proposal", proposalTemplate);
        RegisterTemplate("standard-report", reportTemplate);
    }

    public void RegisterTemplate(string templateId, DocumentBase template)
    {
        _templates[templateId] = template;
        Console.WriteLine($"Template registered: {templateId}");
    }

    public DocumentBase? CreateDocumentFromTemplate(string templateId)
    {
        if (!_templates.TryGetValue(templateId, out var template))
        {
            Console.WriteLine($"Template not found: {templateId}");
            return null;
        }
        return template.Clone();
    }

    public DocumentBase? CreateDocumentFromTemplateDeep(string templateId)
    {
        if (!_templates.TryGetValue(templateId, out var template))
        {
            Console.WriteLine($"Template not found: {templateId}");
            return null;
        }
        return template.DeepClone();
    }

    public void ListAvailableTemplates()
    {
        Console.WriteLine("\nAvailable Templates:");
        Console.WriteLine(new string('=', 50));
        foreach (var template in _templates)
            Console.WriteLine($"ID: {template.Key} - {template.Value.GetDocumentInfo()}");
    }

    public int GetTemplateCount() => _templates.Count;
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/src/services/DocumentTemplateManager.ts
import { DocumentBase } from '../models/abstract/DocumentBase';
import { CVDocument, Education, PersonalInfo } from '../models/concrete/CVDocument';
import { ProposalDocument, ClientInfo, ProposalItem } from '../models/concrete/ProposalDocument';
import { ReportDocument, ReportType, ReportHeader, ReportSection } from '../models/concrete/ReportDocument';

// Prototype Registry: stores pre-configured document templates and produces clones on demand
export class DocumentTemplateManager {
  private readonly _templates = new Map<string, DocumentBase>();

  constructor() {
    this.initializeTemplates();
  }

  private initializeTemplates(): void {
    const cvTemplate = new CVDocument();
    cvTemplate.title = 'Standard CV Template';
    cvTemplate.personalInfo.fullName = '[Your Name]';
    cvTemplate.personalInfo.email = '[your.email@example.com]';
    cvTemplate.personalInfo.phone = '[Your Phone]';
    cvTemplate.personalInfo.address = '[Your Address]';
    cvTemplate.skills.push('[Skill 1]', '[Skill 2]', '[Skill 3]');
    cvTemplate.education.push({
      institution: '[University Name]',
      degree: '[Degree]',
      fieldOfStudy: '[Field of Study]',
      graduationYear: new Date().getFullYear(),
    });

    const proposalTemplate = new ProposalDocument();
    proposalTemplate.title = 'Standard Business Proposal';
    proposalTemplate.clientInfo.companyName = '[Client Company]';
    proposalTemplate.clientInfo.contactPerson = '[Contact Person]';
    proposalTemplate.clientInfo.email = '[client@example.com]';
    proposalTemplate.terms = 'Standard terms and conditions apply.';
    proposalTemplate.validityDays = 30;
    proposalTemplate.proposalItems.push({ description: '[Service/Product Description]', quantity: 1, unitPrice: 0, notes: '' });

    const reportTemplate = new ReportDocument();
    reportTemplate.title = 'Standard Report Template';
    reportTemplate.reportType = ReportType.Management;
    reportTemplate.header.title = '[Report Title]';
    reportTemplate.header.preparedBy = '[Your Name]';
    reportTemplate.header.department = '[Department]';
    reportTemplate.sections.push({ sectionTitle: 'Executive Summary', content: '[Executive summary content]', keyPoints: [], order: 1 });

    this.registerTemplate('standard-cv', cvTemplate);
    this.registerTemplate('standard-proposal', proposalTemplate);
    this.registerTemplate('standard-report', reportTemplate);
  }

  registerTemplate(templateId: string, template: DocumentBase): void {
    this._templates.set(templateId, template);
    console.log(`Template registered: ${templateId}`);
  }

  createDocumentFromTemplate(templateId: string): DocumentBase | null {
    const template = this._templates.get(templateId);
    if (!template) { console.log(`Template not found: ${templateId}`); return null; }
    return template.clone();
  }

  createDocumentFromTemplateDeep(templateId: string): DocumentBase | null {
    const template = this._templates.get(templateId);
    if (!template) { console.log(`Template not found: ${templateId}`); return null; }
    return template.deepClone();
  }

  listAvailableTemplates(): void {
    console.log('\nAvailable Templates:');
    console.log('='.repeat(50));
    for (const [key, template] of this._templates)
      console.log(`ID: ${key} - ${template.getDocumentInfo()}`);
  }

  getTemplateCount(): number { return this._templates.size; }
}
